### PR TITLE
Support external dependencies with Meson

### PR DESCRIPTION
**Pull request by amontoison**
Nick, I have a few questions for you:
- Do we need both Fortran files for the following dummy versions?
```
if not libwsmp.found()
  libgalahad_src += files('wsmp.f', 'wsmp.F90')
endif

if not libpardiso.found()
  libgalahad_src += files('pardiso.f', 'pardiso.F90')
endif
```

- Do we need to compile `spmf_interfaces.F90` and `pastixf_interfaces.F90` for dummy versions of spmf and pastix or for all cases?
```
if not libspmf.found()
  libgalahad_src += files('spmf.F90', 'spmf_interfaces.F90', 'spmf_enums.F90')
endif

if not libpastix.found()
  libgalahad_src += files('pastixf.F90', 'pastixf_interfaces.F90', 'pastixf_enums.F90')
endif
```

- What is the most common library name for the dependency `mkl_pardiso`?

### CODE REVIEW

**Comment by nimgould** [comment]
Hi Alexis

The files wsmp.f and pardiso.f are no longer needed, and I will remove and push.

I believe that we always need the spmf_interfaces and pastix_interfaces files as these are not provided by the pastix people (and really should be for modern fortran).

I believe that mkl_pardiso should be in libmkl_intel_lp64.a or libmkl_intel_lp64.so, but unfortunately these things change. These are currently in i/opt/intel/compilers_and_libraries_2020.1.217/linux/mkl/lib/intel64_lin

**Comment by nimgould** [comment]
Sorry, I meant 
/opt/intel/oneapi/mkl/2023.0.0/lib/intel64
As I hinted, these things change month by month!

**Comment by amontoison** [comment]
Thanks Nick for our answers!
I updated the Meson scripts to handle all dependencies.

I have some comments and new questions:
- We have two different `ssids_ciface.F90`, one in `src/dum/C` and one in `src/ssids/C`. I don't know which file is the good one.
- We have `pastixf_interfaces.F90` and `spmf_interfaces.F90` in `src/dum` but `mkl_paridos_interface.F90` in `src/external/mkl`. We should maybe move all the files related to external interfaces in `src/external`? I also propose to rename the `external/mkl` folder into `external/mkl_pardiso`.
- Do we need to always compile `external/mumps/mumps_types.F90` or only for dummy versions?
- For the dummy version of HSL packages, do we need all these files?
```
if not libhsl.found()
  libgalahad_single_src += files('fa14s.f', 'fd15s.f', 'hsl_ad02s.f90', 'hsl_ma48s.f90','hsl_ma54s.f90', 'hsl_ma57s.f90',
                                 'hsl_ma64s.f90', 'hsl_ma77s.f90', 'hsl_ma86s.f90', 'hsl_ma87s.f90', 'hsl_ma97s.f90',
                                 'hsl_mc34s.f90', 'hsl_mc64s.f90', 'hsl_mc65s.f90', 'ma48s.f', 'mc64s.f', 'mc71s.f',
                                 'hsl_mc68s.f90', 'ma50s.f', 'mc77s.f', 'ma51s.f', 'hsl_mc69s.f90', 'ma57s.f', 'mi21s.f',
                                 'hsl_mc80s.f90', 'ma61s.f', 'hsl_mi20s.f90', 'mi24s.f', 'hsl_mi28s.f90', 'mc13s.f', 'mi26s.f',
                                 'hsl_mi32s.f90', 'mc20s.f', 'hsl_mi35s.f90', 'mc21s.f', 'hsl_of01s.f90', 'mc22s.f', 'mc23s.f',
                                 'hsl_zb01s.f90', 'mc29s.f', 'mc30s.f', 'la04s.f', 'la15s.f', 'mc34s.f', 'ma27s.f', 'mc47s.f',
                                 'mc59s.f', 'zb01s.f', 'mc60s.f', 'mc61s.f', 'ma33s.f')  # 'ma33a.f'

  libgalahad_double_src += files('fa14d.f', 'fd15d.f', 'hsl_ad02d.f90', 'hsl_ma48d.f90', 'hsl_ma54d.f90', 'hsl_ma57d.f90',
                                 'hsl_ma64d.f90', 'hsl_ma77d.f90', 'hsl_ma86d.f90', 'hsl_ma87d.f90', 'hsl_ma97d.f90',
                                 'hsl_mc34d.f90', 'hsl_mc64d.f90', 'hsl_mc65d.f90', 'ma48d.f', 'mc64d.f', 'mc71d.f',
                                 'hsl_mc68d.f90', 'ma50d.f', 'mc77d.f', 'ma51d.f', 'hsl_mc69d.f90', 'ma57d.f', 'mi21d.f',
                                 'hsl_mc80d.f90', 'ma61d.f', 'hsl_mi20d.f90', 'mi24d.f', 'hsl_mi28d.f90', 'mc13d.f', 'mi26d.f',
                                 'hsl_mi32d.f90', 'mc20d.f', 'hsl_mi35d.f90', 'mc21d.f', 'hsl_of01d.f90', 'mc22d.f', 'mc23d.f',
                                 'hsl_zb01d.f90', 'mc29d.f', 'mc30d.f', 'la04d.f', 'la15d.f', 'mc34d.f', 'ma27d.f', 'mc47d.f',
                                 'mc59d.f', 'zb01d.f', 'mc60d.f', 'mc61d.f', 'ma33d.f')  # 'ma33ad.f'

  libgalahad_integer_src += files('hsl_kb22l.f90', 'hsl_mc68i.f90', 'hsl_mc78i.f90', 'hsl_of01i.f90', 'hsl_zb01i.f90', 'kb07i.f')
endif
```
My goal is to have all of them in the `libhsl.so` generated by JuliaHSL but I don't want to add files related to very old packages if they are not used anymore by GALAHAD.
For information, I have some conflicts if I compile both `ma33s.f` / `ma33a.f` or `ma33d.f` / `ma33ad.f`.
I think that we can safely remove the files `ma33a.f` and `ma33ad.f`.

**Comment by amontoison** [comment]
With the last modifications, we only have two shared libraries `libgalahad_single.so` and `libgalahad_double.so` that are generated.

**Comment by amontoison** [comment]
I tested with the shared library and the `.mod` files generated by JuliaHSL and it seems that I should add `ma33` and `mc20` in the JuliaHSL package.
```
/usr/bin/ld : libgalahad_single.so.p/src_gls_gls.F90.o : dans la fonction « __galahad_gls_single_MOD_gls_solve » :
/home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:1615 : référence indéfinie vers « ma33c_ »
/usr/bin/ld : /home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:1639 : référence indéfinie vers « ma33c_ »
/usr/bin/ld : /home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:1648 : référence indéfinie vers « ma33c_ »
/usr/bin/ld : libgalahad_single.so.p/src_gls_gls.F90.o : dans la fonction « __galahad_gls_single_MOD_gls_analyse » :
/home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:863 : référence indéfinie vers « mc20a_ »
/usr/bin/ld : /home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:962 : référence indéfinie vers « ma33a_ »
/usr/bin/ld : /home/alexis/Bureau/git/GALAHAD/builddir/../src/gls/gls.F90:970 : référence indéfinie vers « ma33a_ »
/usr/bin/ld : libgalahad_single.so.p/src_uls_uls.F90.o : dans la fonction « __galahad_uls_single_MOD_uls_initialize_solver » :
/home/alexis/Bureau/git/GALAHAD/builddir/../src/uls/uls.F90:533 : référence indéfinie vers « ma33i_ »
```
What do you think?

**Comment by nimgould** [comment]
Yes, these are needed by gls, which is a fortran 90 version of the hsl archive package ma28; so maybe add ma28 to the juliaHSL package. And if so, probably also ma27 (if that is not already there) as GALAHAD's sls may need that.

**Comment by nimgould** [comment]
For your previous questions (I don't know how to add comments in the right place!) ...

We need both versions of ssids_ciface.F90 as ssids isn't supported with some compilers
(e.g. nagfor). For gfortran/ifort you should use the proper one from ssids/C

Good idea to move those interface files to their proper directories. OK to rename the mkl directory, although we may be using other parts of mkl in the future. I will push these changes.

Yes, we always need to compile external/mumps/mumps_types.F90  as mumps mumps does not provide a standard modern modular fortran interface

As for the HSL files, these will be needed if HSL is not provided. So, not needed for your julia hookup, but possibly for non-julia use.

Thanks for all your attention here

**Comment by nimgould** [comment]
OK, builds now fail because meson build points to mkl not renamed mkl_pardiso. I presume you will fix

**Comment by amontoison** [comment]
Thank you Nick, I fixed the meson build.

- For the dummy HSL routines, you can remove at least `ma33a.f` and `ma33ad.f`. Maybe other ones are also obsolete.
- I tested with `ssids/C/ssids_ciface.F90` and it's not working if I use the dummy versions of ssids (`dum/ssidss.F90` and `dum/ssidsd.F90`).
```
../src/ssids/C/ssids_ciface.F90:32:9:

   32 |     USE SPRAL_SSIDS_types_precision, ONLY : f_ssids_options => ssids_options
      |         1
Fatal Error: Cannot open module file ‘spral_ssids_types_single.mod’ for reading at (1): Aucun fichier ou dossier de ce type
```
I dont have any error if I use `dum/C/ssids_ciface.F90`.
- Bad news: the Meson issue with the module dependencies is back :(

**Comment by amontoison** [comment]
@eli-schwartz
Sorry to annoy you again but our issue with the build order of our Fortran modules is back (https://github.com/ralna/GALAHAD/pull/26#issuecomment-1420011707).
The Meson `depscan.py` is not working because the name of the modules are modified during preprocessing.
I think that your proposition to run the preprocessor on the *.F90 files is a good hack.
It will allow the depscanner to scan the processed file.
What is the best way to do it with Meson?
We know that the option to output the preprocessed file without compilation for our three supported Fortran compilers are:
```
gfortran: -cpp -E
ifort: -fpp -EP
nagfor: -fpp -F
```

Do you also plan to add an `install_modules` function just like `install_headers` in the future?
They are required if we want to link different modern Fortran packages.
The only solution that I found is:
```
install_subdir(meson.project_build_root() / 'libgalahad_double.so.p'
                         install_dir: 'modules',
                         strip_directory: true,
                         exclude_files: ['depscan.dd', 'hsl.dat', 'hsl-deps.json'])
```
but we also have object files and name of the build folder for the library is platform dependent.

**Comment by nimgould** [comment]
Hmm, why can I remove dummy ma33a.f and ma33ad.f? They are needed by gls, and if a user doesn't have HSL they need dummy versions. Or maybe I misunderstood? 

You need to use the dummy version of ssids_ciface.F90 (in dum/C/) if you are using the dummy ssids.F90, you can't mix the two as  SPRAL_SSIDS_types and SPRAL_SSIDS_inform modules will not be available for the C interface case if they aren't available in fortran.

Too bad about the module dependency issue :(

**Comment by amontoison** [comment]
> Hmm, why can I remove dummy ma33a.f and ma33ad.f? They are needed by gls, and if a user doesn't have HSL they need dummy versions. Or maybe I misunderstood?

Because we already have the dummy `ma33s.f` and `ma33d.f` in the `dum` folder.

**Comment by amontoison** [comment]
Nick, can you confirm that `useXXX.F90`, `runXXX_sif.F90` and `inXXX.F90` where `XXX` is the name of the package plus `ptrans.F90` and  `cutest_functions.F90` should be compiled only if we have a `libcustest.so` and if `libcutest.so` is not found we just compile `cutest_dummy.F90`?

**Comment by nimgould** [comment]
Yes, ma33s.f and ma33d.f are in the dum folder, but I still don't understand. Why should I remove these? They aren't anywhere else, and are needed if the proper ma33 is not available.

**Comment by nimgould** [comment]
Yes, the use, run and ptrans files are only for cutest.

**Comment by eli-schwartz** [comment]
> I think that your proposition to run the preprocessor on the *.F90 files is a good hack.
> It will allow the depscanner to scan the processed file.
> What is the best way to do it with Meson?

```meson
real_source = custom_target(
    input: 'foo.F90',
    output: 'foo.F90',
    command: fc.cmd_array() + ['-cpp', '-E', '@INPUT@', '@OUTPUT@'],
)
```

This assumes that you know the correct arguments for the compiler. In meson 0.64.0 we added support for `fc.preprocess('foo.F90', 'bar.F90', 'baz.F90', output: '@PLAINNAME@')` although that's a bit buggy if you invoke it more than once (fixed in meson git master). That will automatically use Meson's own knowledge of all 3 of those compilers (meson claims that nagfor is `-fpp` no -F, and that intel is `-cpp -EP`).



> Do you also plan to add an `install_modules` function just like `install_headers` in the future?

Maybe? I am not sure anyone has proposed it.

> but we also have object files and the name of the build folder for the library is platform dependent.

The object files are indeed a problem, though. The underlying issue, I think, is that fortran doesn't guarantee what the filename of the module file is, does it?

**Comment by amontoison** [comment]
> > I think that your proposition to run the preprocessor on the *.F90 files is a good hack.
> > It will allow the depscanner to scan the processed file.
> > What is the best way to do it with Meson?
> 
> ```meson
> real_source = custom_target(
>     input: 'foo.F90',
>     output: 'foo.F90',
>     command: fc.cmd_array() + ['-cpp', '-E', '@INPUT@', '@OUTPUT@'],
> )
> ```
> 
> This assumes that you know the correct arguments for the compiler. In meson 0.64.0 we added support for `fc.preprocess('foo.F90', 'bar.F90', 'baz.F90', output: '@PLAINNAME@')` although that's a bit buggy if you invoke it more than once (fixed in meson git master). That will automatically use Meson's own knowledge of all 3 of those compilers (meson claims that nagfor is `-fpp` no -F, and that intel is `-cpp -EP`).

Thanks @eli-schwartz! I tried it but Meson returned an error.
I opened an [issue](https://github.com/mesonbuild/meson/issues/11504) in the Meson repository.

> > Do you also plan to add an `install_modules` function just like `install_headers` in the future?
> 
> Maybe? I am not sure anyone has proposed it.

We propose it :) It will very useful for us.
I found this [workaround](https://github.com/dftd4/dftd4/pull/23/files) but it's quite old (2019). I don't think that if it's still relevant.
Other related topics:
- https://github.com/mesonbuild/meson/issues/5374
- https://github.com/mesonbuild/meson/issues/4708

> > but we also have object files and the name of the build folder for the library is platform dependent.
> 
> The object files are indeed a problem, though. The underlying issue, I think, is that fortran doesn't guarantee what the filename of the module file is, does it?

Nick is the Fortran expert, he probably knows the answer.
But is it not possible for Meson to just copy all `*.mod` files from a `lib.private_dir_include()`?

File: meson.build [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+    'warning_level=0',
     'cpp_std=c++17',
   ],
 )
 
 cc = meson.get_compiler('c')
 cxx = meson.get_compiler('cpp')
 fc = meson.get_compiler('fortran')
-fs = import('fs')  # filesystem module
+fs = import('fs')
```
Comment (amontoison) [inline meson.build:18]:
```suggestion
fs = import('fs')

if fc.get_id() == 'gcc'
    add_global_arguments('-cpp', language : 'fortran')
elif fc.get_id() == 'intel' or fc.get_id() == 'nagfor'
    add_global_arguments('-fpp', language : 'fortran')
elif fc.get_id() == 'flang'
    add_global_arguments('-c', language : 'fortran')
endif
```

**Pull request merged**